docs/source/rfc-configuration-overhaul.md

```diff
@@ -0,0 +1,893 @@
+  ENV variables these fields are mapped to.
+
+#### 2. Path to non-existent config file
+
+Providing a path to a configuration file that doesn't exist results in the exact same error as if there was no
+configuration file specified at all.
+
+While it makes sense to silently fallback to ENV variables if the user doesn't provide a path to a config file, in case
+when the user **does** specify the path to a config file, it would be better for the program to fail with an appropriate
+error message.
+
+#### 3. Empty config file
+
+Running Iroha with an empty `config.json` results in the error that does not provide any information about the fallback
+to the environment variables even though it does happen. This behaviour is inconsistent.
+
+```
+Error:
+   0: Please add `PUBLIC_KEY and PRIVATE_KEY` to the configuration.
+```
+
+#### 4. Config file with only `PUBLIC_KEY` and `PRIVATE_KEY` specificed
+
+Running Iroha with a config file that only contains `PUBLIC_KEY` and `PRIVATE_KEY` results in the following error:
+
+```
+Error:
+   0: Please add `p2p_addr` to the configuration.
+```
+
+This error message is misleading and inconsistent:
+
+- If the `p2p_addr` field is added to the config file, the error stays the same.
+- The `p2p_addr` field is not a root-level field, it's actually a part of the `TORII` configuration and should be
+  `TORII.P2P_ADDR`.
+- Error messages asking to add public and private keys use uppercase for configuration fields, while in this error the
+  field name is written in lowercase.
+
+#### 5. Config file with only `PUBLIC_KEY`, `PRIVATE_KEY`, and `TORII.P2P_ADDR` specified
+
+Running Iroha with a config file that only contains `PUBLIC_KEY`, `PRIVATE_KEY`, and `TORII.P2P_ADDR` results in the
+following error:
+
+```
+Error:
+   0: Please add `api_url` to the configuration.
+```
+
+Comparing this scenario to the previous one, we can notice that the `TORII.API_URL` was missing before as well but the
+previous error message didn't mention it.
+
+There are also two inconsistencies:
+
+- Similar to `p2p_addr` in the previous scenario, this field name is lowercase, while for public and private keys it was
+  uppercase.
+- While `TORII.P2P_ADDR` and `TORII.API_URL` share the same root-level (`TORII`), their names are different: "addr"
+  derived from "address" and "URL". Logically these are both addresses or URLs, and it would make sense for the names to
+  be aligned.
+
+#### 6. Config file with invalid `PUBLIC_KEY`
+
+Providing an invalid `PUBLIC_KEY` in the `config.json` results in the same error as when there is no config file at all,
+or an empty config file:
+
+```
+Error:
+   0: Please add `PUBLIC_KEY and PRIVATE_KEY` to the configuration.
+```
+
+This is not a helpful error message as it does not mention what exactly is wrong, why and where the parsing failed.
+
+#### 7. Invalid `IROHA2_PUBLIC_KEY` environment variable
```

Comment:
good research of the all potential cases 👍 wondering if there is other behavior for other combinations of the parameters etc 😱

Reply:
I think that a regular deployment with multiple SDKs doing a request will be helpful.
If (for example) Tuesday's stand-up starts with the automatic news about the `dev`, `stable` or `lts` issues, it'll help both in responding to the community (we know the actual state) and quickly resolving the potential problems.
(I am not sure triggering it only on each release is valid, because Docker may change and influence the behavior of Iroha deployment)